README.md
## SecretColors

[![PyPI version](https://badge.fury.io/py/SecretColors.svg)](https://badge.fury.io/py/SecretColors) [![Documentation Status](https://readthedocs.org/projects/secretcolors/badge/?version=latest)](https://secretcolors.readthedocs.io/en/latest/?badge=latest) 



Library generated for making plots with better color palette. It uses  famous color palettes and adds few helpful functions.   

Currently it supports following Color Palettes

- IBM Color Palette (`ibm`)
- Google Material Design Color Palette (`material`)
- Google Material Design Accent Color Palette (`material-accent`)
- ColorBrewer2 Color Palette (`brewer`)
- VMWare Clarity Color Palette  (`clarity`)
- Tableau Color Palette (`tableau`)

You can get output of colors in variety of color formats including `hex` , `rgb`, `rgba` etc. 

See [changelog](/CHANGELOG.md) to know what's new!

Few sample plots and inspiration behind this library can be found in [WeirdData blog](https://weirddata.github.io/2019/06/11/secret-colors-2.html). 



|<img src="https://user-images.githubusercontent.com/8757115/69130240-72840d00-0ab0-11ea-8ce5-8b715ebef4f6.png" width="60%">|
|:--:|
| Default base colors in matplotlib and SecretColors palettes. |


|<img src="https://user-images.githubusercontent.com/8757115/69130544-fa6a1700-0ab0-11ea-80f4-aaff9e58f804.png" width="60%">|
|:--:|
| Simple bar plot with default colors. |


|<img src="https://user-images.githubusercontent.com/8757115/69130672-31402d00-0ab1-11ea-835d-b161caf4c24a.png" width="60%">|
|:--:|
| Histogram comparison with Magenta and Cyan. You can dramatically change colors by just passing single parameter. |

### Installation 


    pip install SecretColors


### Documentation

Full documentation and API reference can be accessed via [ReadTheDocs](https://secretcolors.readthedocs.io) 

**SecretColors** is a very flexible library. You can easily select different color palettes.

    from SecretColors import Palette
    
    p = Palette()  # Generates Default color palette i.e. IBM Color Palette
    ibm = Palette("ibm")  # Generates IBM Palette
    ibm.red()  # Returns '#fb4b53'
    material = Palette("material")  # Generates Material Palette
    material.red()  # Returns '#f44336'


Select different types of color modes


    p1 = Palette() # Default Color mode (hex)
    p1.green() # '#24a148'
    p2 = Palette(color_mode="hexa")
    p2.green() # '#24a148ff'
    p3 = Palette(color_mode="ahex")
    p3.green() # '#ff24a148'
    p4 = Palette(color_mode="rgb")
    p4.green() # (0.141, 0.631, 0.282)
    p5 = Palette(color_mode="rgba")
    p5.green() # '(0.141, 0.282, 0.631, 1)'


Note: `matplotlib` can accepts *hex*, *rgb* or *hexa* 

Get random colors

    p = Palette()
    p.random() # '#90dbe9'
    p.random(no_of_colors=3) # ['#8fca39', '#64a0fe', '#7430b6']
    p.random(no_of_colors=2, shade=20) # ['#b3e6ff', '#c2dbf4']


Unlimited color manipulations


    p = Palette()
    p.blue()  # normal blue [#408bfc]
    p.blue(shade=20)  # lighter shade of blue [#c9deff]
    p.blue(shade=70)  # darker shade of blue [#054ada]
    p.blue(shade=16.10)  # arbitrary shade of blue (between 0 to 100) [#dbe9ff]
    p.blue(no_of_colors=3)  # Three blue shades ['#b8d4ff', '#408bfc', '#0546d4']
    p.blue(no_of_colors=3, starting_shade=30)  # Three blue shades with lightest one is 30 ['#64a0fe', '#005ef9', '#052ea8']
    p.blue(no_of_colors=3, ending_shade=40)  # Three blue shades with darkest one is 40 ['#edf4ff', '#c9deff', '#97c1ff']
    p.blue(no_of_colors=3, starting_shade=30, ending_shade=40) # Three blue shades with lightest 30 and darkest 60 ['#8cbaff', '#8cbaff', '#8cbaff']
    p.blue(no_of_colors=3, gradient=False) # Three blue shades in random order ['#8cbaff', '#b8d4ff', '#64a0fe']
    p.color_mode = "rgba"
    p.blue(alpha=0.3) # Blue with alpha (0.251, 0.988, 0.545, 0.3) Only works in color mode which outputs alpha values



Flexible functions


    get_complementary("#24a148") # Get complementary color [#a0237c]
    hex_to_rgb("#a0237c") # (0.627, 0.137, 0.486)
    hex_to_hsl("#a0237c") # (0.881, 0.641, 0.382)
    hex_to_hex_a("#a0237c", alpha=0.7) # Get hex with transparency at end [#a0237cb2]
    hex_to_ahex("#a0237c", alpha=0.7) # Get hex with transparency at start [#b2a0237c]
    color_in_between("#24a148","#a0237c") # Color between two ['#616161']
    color_in_between("#24a148","#a0237c", steps=4) # 3 colors between two ['#428154', '#616161', '#80426e'] such that color space is divided into 4 parts
    rgb_to_hex(0.181, 0.241, 0.382) # '#2e3d61'
    rgb_to_hsl(0.181, 0.241, 0.382) # (0.617, 0.357, 0.281)
    hsl_to_hex(0.181, 0.241, 0.382) # '#747849'



Custom and flexible colormaps which can be directly used in `matplotlib` workflow 


    import matplotlib
    import matplotlib.pylab as plt
    import numpy as np
    from SecretColors.cmaps import ColorMap
    from SecretColors import Palette
    p = Palette()
    c = ColorMap(matplotlib, p)
    data = np.random.rand(100, 100)
    plt.pcolor(data, cmap=c.greens())
    plt.show()


Create your own colormaps or make it *qualitative* colormap


    color_list = [p.red(), p.blue()]
    plt.pcolor(data, cmap=c.from_list(color_list, is_qualitative=True))
    plt.show()


Reverse the direction 


    plt.pcolor(data, cmap=c.greens(is_reversed=True))


### TODO
 - [ ] CIE XYZ to CIE Lab conversions
 - [ ] Color Palettes - Latitude, Polaris, Adobe Spectrum
 - [ ] Direct support for more RGB to XYZ conversions
 - [ ] Generate Color Palette from image
 - [x] Color blind safe palette
 - [ ] ~~Out of the box `LinearSegmentedColormap` object~~ 
 
 We can not implement 'out of the box' colormaps for matplotlib because `matplotlib` checks its
  `type` before its use. Hence we can not make new class which can be directly
   act as a substitute.


Note: All CIE-XYZ conversion and colorblind simulation functions are still in
 beta-testing. Do not use them in your production code

### Contribution and Feedback

Feel free to provide feedback and criticism through GitHub or you can email me [rohitsuratekar@gmail.com ](mailto:rohitsuratekar@gmail.com). If you want to contribute, please send pull request to this repository. 

### Acknowledgments

Colors used in this library are partly taken from [IBM Design Language](https://www.ibm.com/design/language/resources/color-library/) , [Google 
Material Design](https://material.io/design/color/the-color-system.html
) , [ColorBrewer](http://colorbrewer2.org/), [VMWare Clarity ](https://vmware.github.io/clarity/documentation/v0.13/color) and 
[Tableau](https://help.tableau.com/current/pro/desktop/en-us/formatting_create_custom_colors.htm#Version_9.x_(legacy)_color_palette_hex_values) .

Color name data is taken from [X11](https://gitlab.freedesktop.org/xorg/app/rgb/raw/master/rgb.txt) 
and [W3]( https://www.w3.org/TR/css-color-3/#svg-color). 

RGB colorspace matrices were taken from [Here](http://www.brucelindbloom.com/index.html?Eqn_RGB_XYZ_Matrix.html) .

Any other specific code or resource used in this library is attributed in
 respective function or method where it is used. It can be accessible
  from our online documentation as well.


### License 

This library and its code is released under MIT License . Read full statement [here](https://github.com/secretBiology/SecretColors/blob/master/LICENSE). 



CHANGELOG.md
# Changelog
# v1.2.6
Removed dependency on `numpy` (Thanks to [ri0t](https://github.com/ri0t))
# v1.2.5
New Color Palette `material-accent` which supports accent colors of Google 
Material Design Color Palette.
# v1.2.3
Added new method to `Palette` class. `Palette.cycle()` will return a 
generator to produce infinite color cycle. Read documentation for 
further 
details. 
# v1.2.2
Fixed bug where `Palette.random()` raising error if `no_of_colors` 
parameter is greater than one.
# v1.2.1
Removed the restrictions from `numpy` version [ see [Issue](https://github.com/secretBiology/SecretColors/issues/4) ]
# v1.2.0
This is one of the major updates since the conception of this project. Many
 changes are essential for robust functionality and proper code maintenance
 . This
 has introduced the following important changes
 * Probably the most important change in this version is return values. Any
  color you generate with this library will return `ColorString` or
   `ColorTuple` (depending on `color_mode`) which are subclasses of `str` and
    `tuple` respectively. This will NOT change anything in your workflow. Check the full documentation for its
      use cases.
 * `SecretColors.palettes` is now `SecretColors.models.palettes.Palette
 `. However, if
  you have used `from SecretColors import palette`, your code will NOT be
   impacted. We recommend importing this way.
 * `ColorMap` is now `SecretColors.cmaps.ColorMap`. You should import it as
  `from SecretColors.cmaps import ColorMap`
 * `Palette.color_between` is now deprecated. Use `utils.color_in_between
 ` instead. Using this function will raise `AttributeError`.
 * Using shade value less than 0 or greater than 100 will throw a `ValueError`.
 * `allow_gray_shades` argument from `Palette` is deprecated and will be
  ignored.
 * `ignore_gray` and `force_gray` arguments from `Palette.random()` are
  deprecated and will be ignored.
 * `starting_shade` and `ending_shade` values (if not provided by user) will
  be set to minimum and maximum shades available in that palette instead 0
   and 100.
 
 ### What's New?
 * New return class `ColorString` and `ColorTuple`
 * New `ColorWheel` for more scientific use
 * New color palette: Tableau
 * Rewritten `ColorMap` class with new usage and easy colormap access to
  famous color palettes.
 * New `Palette.get()` function which supports all X11 and W3 standard names.
 * Many new RGB space conversions (including *XYZ, adobe, srgb, lab* etc)
 * Many new utility methods (check documentation for details)
 * And of course, fixed many bugs!
  
 ### Known Issues
 #### Pycharm type-hinting not working for 'color methods'

We have shifted to custom decorators for dynamically creating
 documentation of many class methods. However, due to this change
 , Pycharm will not show typehint for 'color methods' It is known [bug
 ](https://youtrack.jetbrains.com/issue/PY-30190) in PyCharm regarding type hinting.
 
 #### Converted RGB/CMYK values are not exact
 
 This issue arises when one performs multiple conversion (specially the one
  involving floating points). For example, take a look at following code,
  
 ```python
from SecretColors.utils import rgb_to_hex, hex_to_rgb

r, g, b = 0.3, 0.5, 0.7
hex_color = rgb_to_hex(r, g, b)  # hex_color: #4c80b2
hex_to_rgb(hex_color) # Retuns (0.2980392156862745, 0.5019607843137255, 0.6980392156862745)
```

In the above code, conversion from RGB to HEX and then from HEX to RGB does 
not give us the same output. This issue is very typical of any programming
 language where we are dealing with float calculations. So you should use
   some rounding logic if you want to compare it with original color

In case of conversion to HEX, we usually round the floats to nearest
 integers with default `round` function. 
 
However, our benchmarks and testing suggest that these values are accurate 
with an error of '0.001'. We also check this during our testings. This much 
precision should be good enough for most of the cases. In case, you want even 
better precision, we kindly ask you to implement the method by yourself 
instead of a depending method provided by `SecretColors`. In future, 
we plan to take a look at this in more details. But for now, 
the workaround is to make some rounding function like following
 


 ```python
# Use precision for rounding according to your need
def rounded_rgb(*args):
    return tuple(round(x,2) for x in args)

rounded_rgb(hex_to_rgb(hex_color)) # returns (0.3, 0.5, 0.7)
```

Note: All CIE-XYZ conversion and colorblind simulation functions are still in
 beta-testing. Do not use them in your production code

SecretColors/__colors.py
"""
SecretColors 2019
Author: Rohit Suratekar

Colors are obtained from following sources

IBM Color Palette
(https://www.ibm.com/design/language/elements/color/)
Last Update: 10 April 2019
This contain colors from v2 as well as v1

Material Design Colors
(https://material.io/design/color/#color-theme-creation)

ColorBrewer Color Palette
http://colorbrewer2.org/
Only SingleHue colors are considered here

VMWare Clarity Palette
https://vmware.github.io/clarity/documentation/v0.13/color
Core color is considered as 70
"""

from SecretColors.__color import Color

PALETTE_IBM = "ibm"
PALETTE_MATERIAL = "material"
PALETTE_BREWER = "brewer"
PALETTE_CLARITY = "clarity"

ALL_PALETTES = [PALETTE_IBM, PALETTE_MATERIAL, PALETTE_BREWER, PALETTE_CLARITY]

TYPE_CORE = "core"  # Regular Color
TYPE_GRAY = "gray"  # White, Black and shades of Grey
TYPE_EXTRA = "extra"  # Special palette specific colors

MODE_HEX = "hex"
MODE_RGB = "rgb"
MODE_RGB255 = "rgb255"
MODE_HSL = "hsl"
MODE_RGBA = "rgba"  # With Transparency
MODE_AHEX = "ahex"  # With Transparency
MODE_HSLA = "hsla"  # With Transparency
MODE_HEX_A = "hexa"  # With Transparency


class ParentPalette:
    """
    Base class to create new colors
    """

    def __init__(self, show_warning: bool = True):
        self.show_warning = show_warning

    def get_all_colors(self) -> list:
        """
        :return: List of all colors
        """
        raise NotImplementedError

    def get_palette_name(self) -> str:
        """
        :return: Name of the color palette
        """
        raise NotImplementedError

    def get_creator_url(self) -> str:
        """
        :return: URL of the original resource
        """
        raise NotImplementedError

    def get_shades(self) -> list:
        """
        :return: List of original shades. Note: This will be normalized
        between 0 - 100
        """
        raise NotImplementedError

    def get_core_shade(self) -> int:
        """
        :return: Shade of core color
        """
        raise NotImplementedError

    def get_version(self) -> int:
        """
        :return: Palette version used in this library
        """
        raise NotImplementedError

    def get_last_update(self) -> str:
        """
        :return: Date of last modification happened in this library
        """
        raise NotImplementedError


class IBMPalette(ParentPalette):
    """
    IBM Color Palette
    """

    def get_last_update(self):
        return "10 April 2019"

    def get_version(self):
        return 2

    def get_shades(self):
        return self.shades

    def get_core_shade(self):
        return self.core

    def get_creator_url(self) -> str:
        return "https://www.ibm.com/design/language/elements/color/"

    def get_all_colors(self):
        return [Color(x, self.shades, self.core, self.show_warning) for x in
                self.colors]

    def get_palette_name(self) -> str:
        return "IBM Color Palette"

    shades = [100, 90, 80, 70, 60, 50, 40, 30, 20, 10]
    core = 50
    colors = [
        {
            'n': 'red',
            't': TYPE_CORE,
            'c': ['#2c080a', '#570408', '#750e13', '#a51920', '#da1e28',
                  '#fb4b53', '#ff767c', '#ffa4a9', '#fcd0d3', '#fff0f1']
        },
        {
            'n': 'blue',
            't': TYPE_CORE,
            'c': ['#051243', '#061f80', '#0530ad', '#054ada', '#0062ff',
                  '#408bfc', '#6ea6ff', '#97c1ff', '#c9deff', '#edf4ff']
        },
        {
            'n': 'green',
            't': TYPE_CORE,
            'c': ['#081b09', '#01330f', '#054719', '#10642a', '#198038',
                  '#24a148', '#3dbb61', '#56d679', '#9deeb2', '#dafbe4']
        },
        {
            'n': 'magenta',
            't': TYPE_CORE,
            'c': ['#2a0a16', '#57002b', '#760a3a', '#a11950', '#d12765',
                  '#ee538b', '#fa75a6', '#ffa0c2', '#ffcfe1', '#fff0f6']
        },
        {
            'n': 'cyan',
            't': TYPE_CORE,
            'c': ['#07192b', '#002b50', '#003d73', '#0058a1', '#0072c3',
                  '#1191e6', '#30b0ff', '#6ccaff', '#b3e6ff', '#e3f6ff']
        },
        {
            'n': 'lime',
            't': TYPE_EXTRA,
            'c': ['#1f2a10', '#283912', '#374c1a', '#426200', '#5b8121',
                  '#73a22c', '#81b532', '#95d13c', '#b4e876', '#d7f4bd']
        },

        {
            'n': 'purple',
            't': TYPE_CORE,
            'c': ['#1e1033', '#38146b', '#4f2196', '#6e32c9', '#8a3ffc',
                  '#a66efa', '#bb8eff', '#d0b0ff', '#e6d6ff', '#f7f1ff']
        },

        {
            'n': 'peach',
            't': TYPE_EXTRA,
            'c': ['#3a201b', '#56251a', '#782f1c', '#993a1d', '#c45433',
                  '#fe6100', '#fc835c', '#faad96', '#f8d0c3', '#f7e7e2']
        },

        {
            'n': 'teal',
            't': TYPE_CORE,
            'c': ['#081a1c', '#003137', '#004548', '#006161', '#007d79',
                  '#009c98', '#00bab6', '#20d5d2', '#92eeee', '#dbfbfb']
        },

        {
            'n': 'ultramarine',
            't': TYPE_EXTRA,
            'c': ['#20214f', '#252e6a', '#2e3f8f', '#3151b7', '#3c6df0',
                  '#648fff', '#89a2f6', '#b0bef3', '#d1d7f4', '#e7e9f7']
        },
        {
            'n': 'yellow',
            't': TYPE_CORE,
            'c': ['#372118', '#452f18', '#5b421a', '#70541b', '#91721f',
                  '#b3901f', '#c6a21a', '#e3bc13', '#fed500', '#fbeaae']
        },
        {
            'n': 'cerulean',
            't': TYPE_EXTRA,
            'c': ['#1b2834', '#1d364d', '#1c496d', '#175d8d', '#047cc0',
                  '#009bef', '#56acf2', '#95c4f3', '#c2dbf4', '#deedf7']
        },
        {
            'n': 'aqua',
            't': TYPE_EXTRA,
            'c': ['#122a2e', '#13393e', '#164d56', '#17616b', '#188291',
                  '#12a3b4', '#00b6cb', '#71cddd', '#a0e3f0', '#d1f0f7']
        },

        {
            'n': 'orange',
            't': TYPE_CORE,
            'c': ['#33241c', '#482e1a', '#653d1b', '#814b19', '#ad6418',
                  '#db7c00', '#fe8500', '#fcaf6d', '#fdcfad', '#f5e8de']
        },

        {
            'n': 'indigo',
            't': TYPE_CORE,
            'c': ['#272149', '#352969', '#473793', '#5a3ec8', '#785ef0',
                  '#9b82f3', '#ae97f4', '#c7b6f7', '#dcd4f7', '#e9e8ff']

        },

        {
            'n': 'cool-gray',
            't': TYPE_GRAY,
            'c': ['#13171a', '#242a2e', '#373d42', '#50565b', '#697077',
                  '#868d95', '#9fa5ad', '#b9bfc7', '#d5d9e0', '#f2f4f8']
        },
        {
            'n': 'violet',
            't': TYPE_CORE,
            'c': ['#321c4c', '#44216a', '#602797', '#7732bb', '#9753e1',
                  '#b07ce8', '#bf93eb', '#d2b5f0', '#e2d2f4', '#ece8f5']
        },
        {
            'n': 'gray',
            't': TYPE_GRAY,
            'c': ['#171717', '#282828', '#3d3d3d', '#565656', '#6f6f6f',
                  '#8c8c8c', '#a4a4a4', '#bebebe', '#dcdcdc', '#f3f3f3']
        },
        {
            'n': 'gold',
            't': TYPE_EXTRA,
            'c': ['#2f261c', '#42301b', '#5b421c', '#74521b', '#9c6d1e',
                  '#c4881c', '#e39d14', '#ffb000', '#ffd191', '#f5e8db']
        },
        {
            'n': 'warm-gray',
            't': TYPE_GRAY,
            'c': ['#1a1717', '#2b2828', '#403c3c', '#595555', '#726e6e',
                  '#8f8b8b', '#a7a2a2', '#c1bcbb', '#e0dbda', '#f7f3f1']
        },
        {
            'n': 'white',
            't': TYPE_GRAY,
            'c': ['#ffffff']
        },
        {
            'n': 'black',
            't': TYPE_GRAY,
            'c': ['#000000']
        }

    ]


class MaterialPalette(ParentPalette):
    """
    Material Design Color Palette
    """

    def get_all_colors(self):
        return [Color(x, self.shades, self.core, self.show_warning) for x in
                self.colors]

    def get_palette_name(self):
        return "Material Design Colors"

    def get_creator_url(self):
        return "https://material.io/tools/color"

    def get_shades(self):
        return self.shades

    def get_core_shade(self):
        return self.core

    def get_version(self):
        return 1

    def get_last_update(self):
        return "7 September 2018"

    shades = [900, 800, 700, 600, 500, 400, 300, 200, 100, 50]
    core = 500
    colors = [
        {
            'n': 'red',
            't': TYPE_CORE,
            'c': ['#b71c1c', '#c62828', '#d32f2f', '#e53935', '#f44336',
                  '#ef5350', '#e57373', '#ef9a9a', '#ffcdd2', '#ffebee']
        },
        {
            'n': 'pink',
            't': TYPE_CORE,
            'c': ['#880e4f', '#ad1457', '#c2185b', '#d81b60', '#e91e63',
                  '#ec407a', '#f06292', '#f48fb1', '#f8bbd0', '#fce4ec']
        },
        {
            'n': 'purple',
            't': TYPE_CORE,
            'c': ['#4a148c', '#6a1b9a', '#7b1fa2', '#8e24aa', '#9c27b0',
                  '#ab47bc', '#ba68c8', '#ce93d8', '#e1bee7', '#f3e5f5']
        },
        {
            'n': 'deep-purple',
            't': TYPE_EXTRA,
            'c': ['#311b92', '#4527a0', '#512da8', '#5e35b1', '#673ab7',
                  '#7e57c2', '#9575cd', '#b39ddb', '#d1c4e9', '#ede7f6']
        },
        {
            'n': 'indigo',
            't': TYPE_CORE,
            'c': ['#1a237e', '#283593', '#303f9f', '#3949ab', '#3f51b5',
                  '#5c6bc0', '#7986cb', '#9fa8da', '#c5cae9', '#e8eaf6']
        },
        {
            'n': 'blue',
            't': TYPE_CORE,
            'c': ['#0d47a1', '#1565c0', '#1976d2', '#1e88e5', '#2196f3',
                  '#42a5f5', '#64b5f6', '#90caf9', '#bbdefb', '#e3f2fd']
        },
        {
            'n': 'light-blue',
            't': TYPE_EXTRA,
            'c': ['#01579b', '#0277bd', '#0288d1', '#039be5', '#03a9f4',
                  '#29b6f6', '#4fc3f7', '#81d4fa', '#b3e5fc', '#e1f5fe']
        },
        {
            'n': 'cyan',
            't': TYPE_CORE,
            'c': ['#006064', '#00838f', '#0097a7', '#00acc1', '#00bcd4',
                  '#26c6da', '#4dd0e1', '#80deea', '#b2ebf2', '#e0f7fa']
        },
        {
            'n': 'teal',
            't': TYPE_CORE,
            'c': ['#004d40', '#00695c', '#00796b', '#00897b', '#009688',
                  '#26a69a', '#4db6ac', '#80cbc4', '#b2dfdb', '#e0f2f1']
        },
        {
            'n': 'green',
            't': TYPE_CORE,
            'c': ['#1b5e20', '#2e7d32', '#388e3c', '#43a047', '#4caf50',
                  '#66bb6a', '#81c784', '#a5d6a7', '#c8e6c9', '#e8f5e9']
        },
        {
            'n': 'light-green',
            't': TYPE_EXTRA,
            'c': ['#33691e', '#558b2f', '#689f38', '#7cb342', '#8bc34a',
                  '#9ccc65', '#aed581', '#c5e1a5', '#dcedc8', '#f1f8e9']
        },
        {
            'n': 'lime',
            't': TYPE_EXTRA,
            'c': ['#827717', '#9e9d24', '#afb42b', '#c0ca33', '#cddc39',
                  '#d4e157', '#dce775', '#e6ee9c', '#f0f4c3', '#f9fbe7']
        },
        {
            'n': 'yellow',
            't': TYPE_CORE,
            'c': ['#f57f17', '#f9a825', '#fbc02d', '#fdd835', '#ffeb3b',
                  '#ffee58', '#fff176', '#fff59d', '#fff9c4', '#fffde7']
        },
        {
            'n': 'amber',
            't': TYPE_EXTRA,
            'c': ['#ff6f00', '#ff8f00', '#ffa000', '#ffb300', '#ffc107',
                  '#ffca28', '#ffd54f', '#ffe082', '#ffecb3', '#fff8e1']
        },
        {
            'n': 'orange',
            't': TYPE_CORE,
            'c': ['#e65100', '#ef6c00', '#f57c00', '#fb8c00', '#ff9800',
                  '#ffa726', '#ffb74d', '#ffcc80', '#ffe0b2', '#fff3e0']
        },
        {
            'n': 'deep-orange',
            't': TYPE_EXTRA,
            'c': ['#bf360c', '#d84315', '#e64a19', '#f4511e', '#ff5722',
                  '#ff7043', '#ff8a65', '#ffab91', '#ffccbc', '#fbe9e7']
        },
        {
            'n': 'brown',
            't': TYPE_CORE,
            'c': ['#3e2723', '#4e342e', '#5d4037', '#6d4c41', '#795548',
                  '#8d6e63', '#a1887f', '#bcaaa4', '#d7ccc8', '#efebe9']
        },
        {
            'n': 'gray',
            't': TYPE_GRAY,
            'c': ['#212121', '#424242', '#616161', '#757575', '#9e9e9e',
                  '#bdbdbd', '#e0e0e0', '#eeeeee', '#f5f5f5', '#fafafa']
        },
        {
            'n': 'blue-gray',
            't': TYPE_GRAY,
            'c': ['#263238', '#37474f', '#455a64', '#546e7a', '#607d8b',
                  '#78909c', '#90a4ae', '#b0bec5', '#cfd8dc', '#eceff1']
        },
        {
            'n': 'black',
            't': TYPE_GRAY,
            'c': ['#000000']
        },
        {
            'n': 'white',
            't': TYPE_GRAY,
            'c': ['#ffffff']
        }

    ]


class ColorBrewer(ParentPalette):
    """
    Color Brewer Color Paletter
    """

    def get_all_colors(self) -> list:
        return [Color(x, self.shades, self.core, self.show_warning) for x in
                self.colors]

    def get_palette_name(self) -> str:
        return "Color Brewer Color Palette"

    def get_creator_url(self) -> str:
        return "http://colorbrewer2.org/"

    def get_shades(self) -> list:
        return self.shades

    def get_core_shade(self) -> int:
        return self.core

    def get_version(self) -> int:
        return 1

    def get_last_update(self) -> str:
        return "12 April 2019"

    shades = [90, 80, 70, 60, 50, 40, 30, 20, 10]
    core = 60
    colors = [
        {
            'n': 'blue',
            't': TYPE_CORE,
            'c': ['#08306b', '#08519c', '#2171b5', '#4292c6', '#6baed6',
                  '#9ecae1', '#c6dbef', '#deebf7', '#f7fbff']
        },
        {
            'n': 'purple',
            't': TYPE_CORE,
            'c': ['#3f007d', '#54278f', '#6a51a3', '#807dba', '#9e9ac8',
                  '#bcbddc', '#dadaeb', '#efedf5', '#fcfbfd']
        },
        {
            'n': 'green',
            't': TYPE_CORE,
            'c': ['#00441b', '#006d2c', '#238b45', '#41ab5d', '#74c476',
                  '#a1d99b', '#c7e9c0', '#e5f5e0', '#f7fcf5']
        },
        {
            'n': 'orange',
            't': TYPE_CORE,
            'c': ['#7f2704', '#a63603', '#d94801', '#f16913', '#fd8d3c',
                  '#fdae6b', '#fdd0a2', '#fee6ce', '#fff5eb']
        },
        {
            'n': 'red',
            't': TYPE_CORE,
            'c': ['#67000d', '#a50f15', '#cb181d', '#ef3b2c', '#fb6a4a',
                  '#fc9272', '#fcbba1', '#fee0d2', '#fff5f0']
        },
        {
            'n': 'gray',
            't': TYPE_GRAY,
            'c': ['#000000', '#252525', '#525252', '#737373', '#969696',
                  '#bdbdbd', '#d9d9d9', '#f0f0f0', '#ffffff']
        },
        {
            'n': 'black',
            't': TYPE_GRAY,
            'c': ['#000000']
        },
        {
            'n': 'white',
            't': TYPE_GRAY,
            'c': ['#ffffff']
        }
    ]


class ClarityPalette(ParentPalette):
    def get_all_colors(self) -> list:
        return [Color(x, self.shades, self.core, self.show_warning) for x in
                self.colors]

    def get_palette_name(self) -> str:
        return "VMWare Clarity Color Palette"

    def get_creator_url(self) -> str:
        return "https://vmware.github.io/clarity/documentation/v0.13/color"

    def get_shades(self) -> list:
        return self.shades

    def get_core_shade(self) -> int:
        return self.core

    def get_version(self) -> int:
        return 1

    def get_last_update(self) -> str:
        return "12 April 2019"

    shades = [100, 90, 80, 70, 60, 50, 40, 30, 20, 10]
    core = 70
    colors = [
        {
            'n': 'red',
            't': TYPE_CORE,
            'c': ['#a32100', '#c92100', '#e12200', '#f52f22', '#f54f47',
                  '#f76f6c', '#f89997', '#f8b7b6', '#f5dbd9', '#fff0ee']
        },
        {
            'n': 'pink',
            't': TYPE_CORE,
            'c': ['#9b0d54', '#b0105b', '#c41261', '#d91468', '#ed186f',
                  '#f1428a', '#f46ca5', '#f897bf', '#fbc1da', '#ffebf5']
        },
        {
            'n': 'purple',
            't': TYPE_CORE,
            'c': ['#4d007a', '#660092', '#781da0', '#8939ad', '#9b56bb',
                  '#ad73c8', '#be90d6', '#d0ace4', '#e1c9f1', '#f3e6ff']
        },
        {
            'n': 'ultramarine',
            't': TYPE_EXTRA,
            'c': ['#0f1e82', '#1a23a0', '#343dac', '#4e56b8', '#6870c4',
                  '#838acf', '#9da3db', '#b7bde7', '#d1d6f3', '#ebf0ff']
        },
        {
            'n': 'blue',
            't': TYPE_CORE,
            'c': ['#003d79', '#004d8a', '#0065ab', '#0079b8', '#0095d3',
                  '#49afd9', '#89cbdf', '#a6d8e7', '#c5e5ef', '#e1f1f6']
        },
        {
            'n': 'cyan',
            't': TYPE_CORE,
            'c': ['#004a70', '#005680', '#006690', '#0081a7', '#009cbf',
                  '#00b7d6', '#36c9e1', '#6ddbeb', '#a3edf6', '#ccfbff']
        },
        {
            'n': 'teal',
            't': TYPE_CORE,
            'c': ['#006668', '#007e7a', '#00968b', '#00ab9a', '#00bfa9',
                  '#00d4b8', '#38dfc8', '#6fead9', '#a7f4e9', '#defff9']
        },
        {
            'n': 'green',
            't': TYPE_CORE,
            'c': ['#1d5100', '#266900', '#2f8400', '#48960c', '#62a420',
                  '#60b515', '#85c81a', '#aadb1e', '#c7e59c', '#dff0d0']
        },
        {
            'n': 'yellow',
            't': TYPE_CORE,
            'c': ['#c47d00', '#d28f00', '#dfa100', '#edb200', '#fac400',
                  '#fdd006', '#ffdc0b', '#ffe860', '#fef3b5', '#fffce8']
        },
        {
            'n': 'orange',
            't': TYPE_CORE,
            'c': ['#aa4500', '#c25400', '#d36000', '#e46c00', '#f57600',
                  '#ff8400', '#ff9c32', '#ffb565', '#ffcd97', '#ffe5c9']
        },
        {
            'n': 'red-orange',
            't': TYPE_EXTRA,
            'c': ['#cd3517', '#de400f', '#ee4a08', '#ff5500', '#ff681c',
                  '#ff8142', '#ff9a69', '#ffb38f', '#ffccb5', '#ffe5dc']
        },
        {
            'n': 'warm-gray',
            't': TYPE_GRAY,
            'c': ['#5b4d47', '#6c5f59', '#80746d', '#948981', '#a89e95',
                  '#bbb3a9', '#cfc8bd', '#e3ddd1', '#f4f1e6', '#faf9f5']
        },
        {
            'n': 'neutral-gray',
            't': TYPE_GRAY,
            'c': ['#313131', '#444444', '#565656', '#737373', '#9a9a9a',
                  '#cccccc', '#dddddd', '#eeeeee', '#f2f2f2', '#fafafa']
        },
        {
            'n': 'cool-gray',
            't': TYPE_GRAY,
            'c': ['#25333d', '#314351', '#495a67', '#61717d', '#798893',
                  '#919fa8', '#a9b6be', '#c1cdd4', '#d9e4ea', '#f3f6fa']
        },
        {
            'n': 'black',
            't': TYPE_GRAY,
            'c': ['#000000']
        },
        {
            'n': 'white',
            't': TYPE_GRAY,
            'c': ['#ffffff']
        }
    ]
